lib/maths/CBoostedTreeFactory.cc

```diff
@@ -326,70 +350,117 @@ void CBoostedTreeFactory::initializeHyperparameters(core::CDataFrame& frame) {
     }
 }
 
 void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDataFrame& frame) {
 
     // The strategy here is to:
     //   1) Get percentile estimates of the gain in the loss function and its sum
-    //      curvature from the splits selected in a single tree with regulizers
+    //      curvature from the splits selected in a single tree with regularisers
     //      zeroed,
-    //   2) Use these to upper bound the size of alpha, gamma and lambda, that is
-    //      find values we expect to underfit the data,
-    //   3) Decrease each regularizer and look for a turning point in the test
-    //      loss, i.e. the point at which transition to overfit occurs.
+    //   2) Use these to extract reasonable intervals to search for the multipliers
+    //      for the various regularisation penalties,
+    //   3) Line search these intervales for a turning point in the test loss for
+    //      the base learner, i.e. the point at which transition to overfit occurs.
+    //
     // We'll search intervals in the vicinity of these values in the hyperparameter
     // optimisation loop.
 
-    core::CPackedBitVector allTrainingRowsMask{m_TreeImpl->allTrainingRowsMask()};
-
-    double log2MaxTreeSize{std::log2(
-        static_cast<double>(m_TreeImpl->maximumTreeSize(allTrainingRowsMask)))};
+    double log2MaxTreeSize{std::log2(static_cast<double>(m_TreeImpl->maximumTreeSize(
+                               m_TreeImpl->m_TrainingRowMasks[0]))) +
+                           1.0};
     m_TreeImpl->m_Regularization.softTreeDepthLimit(
         m_TreeImpl->m_RegularizationOverride.softTreeDepthLimit().value_or(log2MaxTreeSize));
     m_TreeImpl->m_Regularization.softTreeDepthTolerance(
         m_TreeImpl->m_RegularizationOverride.softTreeDepthTolerance().value_or(
             0.5 * (MIN_SOFT_DEPTH_LIMIT_TOLERANCE + MAX_SOFT_DEPTH_LIMIT_TOLERANCE)));
     LOG_TRACE(<< "max depth = " << m_TreeImpl->m_Regularization.softTreeDepthLimit()
-              << ", tol = " << m_TreeImpl->m_Regularization.softTreeDepthTolerance());
+              << ", tolerance = " << m_TreeImpl->m_Regularization.softTreeDepthTolerance());
+
+    auto gainAndTotalCurvaturePerNode =
+        this->estimateTreeGainAndCurvature(frame, {1.0, 50.0, 90.0});
+    LOG_TRACE(<< "gains and total curvatures per node = "
+              << core::CContainerPrinter::print(gainAndTotalCurvaturePerNode));
 
-    double gainPerNode;
-    double totalCurvaturePerNode;
-    std::tie(gainPerNode, totalCurvaturePerNode) =
-        this->estimateTreeGainAndCurvature(frame, allTrainingRowsMask);
-    LOG_TRACE(<< "gain per node = " << gainPerNode
-              << " total curvature per node = " << totalCurvaturePerNode);
+    double gainPerNode1stPercentile{gainAndTotalCurvaturePerNode[0].first};
+    double gainPerNode50thPercentile{gainAndTotalCurvaturePerNode[1].first};
+    double gainPerNode90thPercentile{gainAndTotalCurvaturePerNode[2].first};
```

Comment:
So this was partly an empirical observation, in fact I started with a symmetric interval. At 90th percentile we severely curtail the tree, i.e. in all the data sets I tested this majorly under fit, so using a higher value seemed unnecessary. By contrast, we could sometimes still eek more performance out if I cut this off too high. My hand wavy argument would be that since this applies to the base learner we should be overfitting somewhat, so we should prefer to explore smaller values.

Reply:
Ok. I am satisfied with your hand wavy argument :)

Reply:
Curious: What were the characteristics of the datasets that were used to derive this empirical observation? Would it be possible to take a look at these?